third_party/pocketfft/BUILD.bazel

```diff
@@ -4,8 +4,21 @@ package(default_visibility = ["//visibility:public"])
 
 cc_library(
     name = "pocketfft",
-    hdrs = ["pocketfft_hdronly.h"],
-    copts = ["-fexceptions"],
+    hdrs = ["src/ducc0/fft/fft.h"],
+    srcs = ["src/ducc0/fft/fft1d.h",
+            "src/ducc0/infra/aligned_array.h",
+            "src/ducc0/infra/error_handling.h",
+            "src/ducc0/infra/mav.h",
+            "src/ducc0/infra/simd.h",
+          "src/ducc0/infra/threading.h",
+          "src/ducc0/infra/threading.cc",
+            "src/ducc0/infra/useful_macros.h",
+            "src/ducc0/math/cmplx.h",
+            "src/ducc0/math/unity_roots.h",
+],
+    copts = ["-fexceptions", "-ffast-math"],
```

Comment:
`-ffast-math`(or at least a subset of it) does indeed have a large impact, because without this flag the compiler is not allowed to use FMA instructions ... and FFTs consist of mostly FMAs. I know that there is a `#pragma` which is supposed to allow this, but `gcc` has a history of not dealing with this properly (I don't know about `clang`).
My idea would be to switch this on only for the  compilation of `jaxlib/pocketfft_kernels.cc`, if that's possible.

Reply:
Maybe we can get away with `-ffp-contract=fast` I'd have to test that with several compilers though. In any case this is the most important component of `-ffast-math` for this code.

Reply:
Interesting.

For clang to use fma instructions I think you need two things:
a) to target an architecture with FMA instructions, and
b) to enable `fp-contract`: https://clang.llvm.org/docs/UsersManual.html#cmdoption-ffp-contract
I assume gcc is similar.

(b) is no problem: we can just turn that flag on for this file: it's just a slightly more nuanced `-ffast-math`.

However, for (a) we don't distribute our Python wheels built for an architecture that has FMA instructions. We currently assume AVX but not AVX2 or FMA. So we'd want to start building for, say, Haswell machines, which introduced AVX2 and FMA, to get the full benefit.

(Function multi-versioning is technically possible but it's problematic in C++ code because it runs a risk of introducing [ODR problems](https://en.wikipedia.org/wiki/One_Definition_Rule).)

I think that's doable, but certainly even the AVX requirement has sometimes tripped people up! Apparently plenty of machines from older than 2013 are still out there!

Reply:
If I remember correctly, I also encountered some situations where `gcc` refused to vectorize certain expressions without `-ffast-math`; if that is true, it could have an effect on pure AVX architectures (without FMA) as well. But this is probably more a concern for the other `ducc` components like spherical harmonic transforms, and not so much for FFT.
Let's see how far we get without the sledgehammer `-ffast-math` :-)